test-hydration/test/Hydration.purs

```diff
@@ -0,0 +1,119 @@
+  case mRootElement of
+    Just rootElement →
+      pure $ Element.toNode rootElement
+    Nothing →
+      throwError $ error "Cannot find #root in document."
+
+renderHydrate
+  ∷ ∀ m a
+  . MonadEffect m
+  ⇒ MonadThrow Error m
+  ⇒ (Boolean → Basic.VDom a)
+  → m Unit
+renderHydrate renderFn = liftEffect do
+  let
+    serverHTML ∷ String
+    serverHTML = "<!DOCTYPE html><html><head></head><body>"
+      <> Render.render (renderFn false)
+      <> "</body></html>"
+
+  jsdom ← JSDOM.make serverHTML
+  window ← JSDOM.window jsdom
+  htmlDocument ← Window.document window
+  initialNode ← findInitialNode htmlDocument
+
+  let
+    document ∷ DOM.Document
+    document = HTMLDocument.toDocument htmlDocument
+
+    driverSpec ∷ Driver.DriverSpec Boolean a
+    driverSpec = { initialState: true, render: renderFn }
+
+  void $ Driver.hydrateUI document initialNode driverSpec
+
+main ∷ Effect Unit
+main = launchAff_ $ runSpec [ consoleReporter ] do
+  describe "Hydration" do
+    it "works" do
+      let
+        general ∷ Boolean → Basic.VDom Void
+        general _ =
+          Basic.elem "div" [ "className" .= "container", "id" := "root" ]
+            [ Basic.elem "label" [ "htmlFor" .= "username" ]
+                [ Basic.text "Username" ]
+            , Basic.elem "input" [ "id" := "input" ] []
+            , Basic.elem "a" [ "href" := "index" ] [ Basic.text "Inbox" ]
+            , Basic.keyed "div" []
+                [ "0" /\ Basic.elem "span" [] [ Basic.text "0" ]
+                , "1" /\ Basic.elem "span" [] [ Basic.text "1" ]
+                ]
+            ]
+      renderHydrate general `shouldReturn` unit
+
+    it "fails on mismatched properties" do
+      let
+        mismatch ∷ Boolean → Basic.VDom Void
+        mismatch isClient =
+          Basic.elem "div" [ "className" .= show isClient, "id" := "root" ] []
+      expectError $ renderHydrate mismatch
+
+    it "fails on mismatched attributes" do
+      let
+        mismatch ∷ Boolean → Basic.VDom Void
+        mismatch isClient =
+          Basic.elem "div" [ "style" := show isClient, "id" := "root" ] []
+      expectError $ renderHydrate mismatch
+
+    it "fails on mismatched elements" do
+      let
+        mismatch ∷ Boolean → Basic.VDom Void
+        mismatch isClient =
+          Basic.elem (if isClient then "div" else "span") [ "id" := "root" ] []
+      expectError $ renderHydrate mismatch
+
+    it "fails on mismatched text nodes" do
+      let
+        mismatch ∷ Boolean → Basic.VDom Void
+        mismatch isClient =
+          Basic.elem "div" [ "id" := "root" ] [ Basic.text $ show isClient ]
+      expectError $ renderHydrate mismatch
+
+    it "fails on mismatched children" do
+      let
+        mismatch ∷ Boolean → Basic.VDom Void
+        mismatch isClient =
+          Basic.elem "div" [ "id" := "root" ] $
+            if isClient then []
+            else [ Basic.elem "span" [] [ Basic.text "Server!" ] ]
+      expectError $ renderHydrate mismatch
```

Comment:
I would also add a test

1. if trying to hydrate `Basic.elem "div" [ "] [ Basic.text "1", Basic.text "2" ]` on `<div>12</div>` then it will not throw error (bc it should split `TEXT_NODE "12"` on `TEXT_NODE "1"` and  `TEXT_NODE "2"`
2. if trying to hydrate `Basic.elem "div" [ "] [ Basic.text "1", Basic.text "" ]` on `<div>1</div>` then it will not throw error (bc it should add to `TEXT_NODE "1"` the  `TEXT_NODE ""`